deploy/lib/xquery/setup.xqy

```diff
@@ -1067,24 +1244,93 @@ declare function setup:validate-mimetypes($import-config as element(configuratio
 };
 
-declare function setup:create-forests($import-config as element(configuration)) as item()*
+(:
+  Initialize the scale-out cleanup state files based on the latest bootstrap request.
+:)
+declare function setup:initialize-cleanup-state( $import-config as element(configuration), $do-internals as xs:boolean )
 {
-  for $db-config in setup:get-databases-from-config($import-config)
-  let $database-name := setup:get-database-name-from-database-config($db-config)
-  let $forests-per-host as xs:integer? := $db-config/db:forests-per-host
-  where fn:not($database-name = 'filesystem')
+  (: Get which files to use, based on whether using internals :)
+  let $which-delete-map-file := 
+        if( $do-internals ) then
+          $delete-map-file-internal
+        else
+          $delete-map-file
+  let $which-replicating-map-file := 
+        if( $do-internals ) then
+          $replicating-map-file-internal
+        else
+          $replicating-map-file
+
   return
-    if (fn:exists($forests-per-host)) then
-      setup:create-forests-from-count($import-config, $db-config, $database-name, $forests-per-host)
-    else
-      setup:create-forests-from-config($import-config, $db-config, $database-name)
+    (: Map of forests to delete must exist - otherwise, why bother :)
+    if( fn:doc-available( $which-delete-map-file ) ) then
+      let $local-delete-map := map:map( fn:doc( $which-delete-map-file )/node() )
+      let $local-replicating-map := map:map( fn:doc( $which-replicating-map-file )/node() )
+
+      let $_ := xdmp:log( "INIT delete map file: " || $which-delete-map-file )
+      let $_ := for $i in map:keys( $local-delete-map )
+        let $_ := map:put( $delete-map, $i, map:get( $local-delete-map, $i ) )
+        return xdmp:log( "     -> " || $i || " : " || map:get( $delete-map, $i ) )
+
+      let $_ := xdmp:log( "INIT replicating map file: " || $which-replicating-map-file )
+      let $_ := for $i in map:keys( $local-replicating-map)
+        let $_ := map:put( $replicating-map, $i, map:get( $local-replicating-map, $i ) )
+        return xdmp:log( "     -> " || $i || " : " || map:get( $replicating-map, $i ) )
+
+      return()
+    else ()
+};
+
+(:
+  Save the scale-out cleanup state files to the database.  These are later used on subsequent runs of roxy
+  when a "clean replicas" is issued.
+:)
+declare function setup:save-cleanup-state( $import-config as element(configuration), $do-internals as xs:boolean )
+{
+  let $which-delete-map-file := 
+        if( $do-internals ) then
+          $delete-map-file-internal
+        else
+          $delete-map-file
+  let $which-replicating-map-file := 
+        if( $do-internals ) then
+          $replicating-map-file-internal
+        else
+          $replicating-map-file
+
+  (: Write the delete maps and the replicating maps for use when delete old replicas is done :)
+  return
+  (
+    xdmp:document-insert( $which-delete-map-file, document { $delete-map } ),
```

Comment:
Could you explicitly force admin permissions on these to prevent users from accidentally deleting them?